include/fiction/algorithms/simulation/sidb/determine_displacement_robustness.hpp

```diff
@@ -0,0 +1,542 @@
+    /**
+     * Cells/SiDBs of the layout which are not affected by variations.
+     */
+    std::set<cell<Lyt>> fixed_cells{};
+    /**
+     * This flag controls whether the displacement in the y-direction can lead to changes in the Si-dimer.
+     */
+    bool allow_dimer_change_in_y_direction = false;
+};
+/**
+ * Statistics for the displacement robustness domain computation.
+ */
+struct displacement_robustness_domain_stats
+{
+    /**
+     * Total runtime in sec. to determine the robustness of the SiDB gate.
+     */
+    mockturtle::stopwatch<>::duration time_total{0};
+    /**
+     * The number of operational SiDB gates resulting from the given layout by displacements.
+     */
+    std::size_t num_operational_sidb_displacements{0};
+    /**
+     * The number of non-operational SiDB gates resulting from the given layout by displacements.
+     */
+    std::size_t num_non_operational_sidb_displacements{0};
+};
+
+namespace detail
+{
+
+template <typename Lyt, typename TT>
+class displacement_robustness_domain_impl
+{
+  public:
+    /**
+     * Standard constructor. Initializes the layout, the truth table, the parameters, and the statistics.
+     *
+     * @param lyt SiDB cell-level layout to be evaluated regarding displacement robustness.
+     * @param spec Expected Boolean function of the layout given as a multi-output truth table.
+     * @param ps Parameters for the displacement robustness computation.
+     * @param st Statistics related to the displacement robustness computation.
+     */
+    displacement_robustness_domain_impl(const Lyt& lyt, const std::vector<TT>& spec,
+                                        const displacement_robustness_domain_params<Lyt>& ps,
+                                        displacement_robustness_domain_stats&             st) noexcept :
+            layout{lyt},
+            params{ps},
+            stats{st},
+            truth_table{spec},
+            generator(rd())
+    {
+        for ([[maybe_unused]] const auto& c : params.fixed_cells)
+        {
+            assert(!layout.is_empty_cell(c) && "Not all fixed cells are part of the layout");
+        }
+
+        assert(
+            (is_operational(layout, truth_table, params.operational_params).first == operational_status::OPERATIONAL) &&
+            "The given layout is not a valid SiDB gate implementation");
+
+        sidbs_of_the_original_layout.reserve(layout.num_cells());
+        layout.foreach_cell([&](const auto& c) { sidbs_of_the_original_layout.push_back(c); });
+        calculate_permitted_displacements_for_each_sidb();
+    };
+    /**
+     * This function calculates the robustness domain of the SiDB gate based on the provided truth table specification
+     * and displacement robustness computation parameters.
+     */
+    displacement_robustness_domain<Lyt> determine_robustness_domain()
+    {
+        mockturtle::stopwatch stop{stats.time_total};
+        calculate_permitted_displacements_for_each_sidb();
```

Comment:
This function is already called in the constructor. Why is it necessary to call it twice?

Reply:
good point.